src/goLanguageServer.ts

```diff
@@ -379,3 +393,96 @@ function registerUsualProviders(ctx: vscode.ExtensionContext) {
 	vscode.workspace.onDidChangeTextDocument(parseLiveFile, null, ctx.subscriptions);
 }
+
+async function shouldUpdateLanguageServer(tool: Tool, path: string): Promise<boolean> {
+	// Only support updating gopls for now.
+	if (tool.name !== 'gopls') {
+		return false;
+	}
+	// First, run the "gopls version" command and parse its results.
+	// If "gopls" is so old that it doesn't have the "gopls version" command,
+	// or its version doesn't match our expectations, prompt the user to download.
+	const usersVersion = await goplsVersion(path);
+	if (!usersVersion) {
+		return true;
+	}
+	// We might have a developer version. Don't make the user update.
+	if (usersVersion === '(devel)') {
+		return false;
+	}
+	// If the user has a version of gopls that we understand,
+	// ask the proxy for the latest version, and if the user's version is older,
+	// prompt them to update.
+	const listURL = `https://proxy.golang.org/${tool.importPath}/@v/list`;
+	const data = await WebRequest.json<string>(listURL);
+
+	// Coerce the versions into SemVers so that they can be sorted correctly.
+	const versions = [];
+	for (const version of data.trim().split(os.EOL)) {
+		versions.push(semver.coerce(version));
+	}
+	if (versions.length === 0) {
+		return false;
+	}
+	versions.sort();
+	const latestVersion = versions[versions.length - 1];
+	return semver.lt(usersVersion, latestVersion);
+}
+
+async function goplsVersion(goplsPath: string): Promise<string> {
+	const env = getToolsEnvVars();
+	const util = require('util');
+	const execFile = util.promisify(cp.execFile);
+	let output: any;
+	try {
+		const { stdout } = await execFile(goplsPath, ['version'], { env });
+		output = stdout;
+	} catch (e) {
+		// The "gopls version" command is not supported, or something else went wrong.
+		// TODO: Should we propagate this error?
+		return null;
+	}
+
+	const lines = <string>output.trim().split(os.EOL);
+	switch (lines.length) {
+		case 0:
+			// No results, should update.
+			// Worth doing anything here?
+			return null;
+		case 1:
+			// Built in $GOPATH mode. Should update.
+			// TODO: Should we check the Go version here?
+			// Do we even allow users to enable gopls if their Go version is too low?
+			return null;
+		case 2:
+			// We might actually have a parseable version.
+			break;
+		default:
+			return null;
+	}
+
+	// The second line should be the sum line.
+	// It should look something like this:
+	//
+	//    golang.org/x/tools/gopls@v0.1.3 h1:CB5ECiPysqZrwxcyRjN+exyZpY0gODTZvNiqQi3lpeo=
+	//
+	// TODO: We should use a regex to match this, but for now, we split on the @ symbol.
+	// The reasoning for this is that gopls still has a golang.org/x/tools/cmd/gopls binary,
+	// so users may have a developer version that looks like "golang.org/x/tools@(devel)".
+	const moduleVersion = lines[1].trim().split(' ')[0];
+
+	// Get the relevant portion, that is:
+	//
+	//    golang.org/x/tools/gopls@v0.1.3
+	//
+	const split = moduleVersion.trim().split('@');
```

Comment:
`moduleVersion` can be `undefined` if `lines[1]` had no `' '`

Reply:
I don't believe it would ever be. I tested a few cases, and in the case where there is no element to split on, split just returns an array of 1 element containing the original string.